src/coreclr/nativeaot/Runtime/DebugHeader.cpp

```diff
@@ -0,0 +1,275 @@
+//     be used in the manner the format specification describes.
+struct NativeAOTRuntimeDebugHeader
+{
+    // The cookie serves as a sanity check against process corruption or being requested
+    // to treat some other non-.Net module as though it did contain the coreRT runtime.
+    // It can also be changed if we want to make a breaking change so drastic that
+    // earlier debuggers should treat the module as if it had no .Net runtime at all.
+    // If the cookie is valid a debugger is safe to assume the Major/Minor version fields
+    // will follow, but any contents beyond that depends on the version values.
+    // The cookie value is currently set to 0x4E 0x41 0x44 0x48 (NADH in ascii)
+    const uint8_t Cookie[4] = { 0x4E, 0x41, 0x44, 0x48 };
+
+    // This counter can be incremented to indicate breaking changes
+    // This field must be encoded little endian, regardless of the typical endianess of
+    // the machine
+    const uint16_t MajorVersion = 1;
+
+    // This counter can be incremented to indicate back-compatible changes
+    // This field must be encoded little endian, regardless of the typical endianess of
+    // the machine
+    const uint16_t MinorVersion = 0;
+
+    // These flags must be encoded little endian, regardless of the typical endianess of
+    // the machine. Ie Bit 0 is the least significant bit of the first byte.
+    // Bit 0 - Set if the pointer size is 8 bytes, otherwise pointer size is 4 bytes
+    // Bit 1 - Set if the machine is big endian
+    // The high 30 bits are reserved. Changes to these bits will be considered a
+    // back-compatible change.
+    const uint32_t Flags = sizeof(void*) == 8 ? 0x1 : 0x0;
+
+    // Reserved - Currently it only serves as alignment padding for the pointers which
+    // follow but future usage will be considered a back-compatible change.
+    const uint32_t ReservedPadding = 0;
+
+    // Header pointers below here are encoded using the defined pointer size and endianess
+    // specified in the Flags field. The data within the contracts they point to also uses
+    // the same pointer size and endianess encoding unless otherwise specified.
+
+    DebugTypeEntry *DebugTypesList = nullptr;
+
+    GlobalValueEntry *GlobalsList = nullptr;
+
+    DefineEntry *DefinesList = nullptr;
+};
+
+extern "C" NativeAOTRuntimeDebugHeader g_NativeAOTRuntimeDebugHeader = {};
+
+#define MAKE_DEBUG_ENTRY(TypeName, FieldName, Value)                                                                        \
+    do                                                                                                                      \
+    {                                                                                                                       \
+        currentType = new (nothrow) DebugTypeEntry{ previousType, #TypeName, #FieldName, Value };                             \
```

Comment:
I'm wondering if we could pre-initialize most of these in the data region. The extra `new`s are probably not measurably expensive, but if we can avoid doing doing extra work at startup, it's better to avoid it.

Reply:
+1 I was about to write the same feedback.

Reply:
I originally intended to have it be all at compile time. The data is just offsets and build defines so it is available at  compile time.  I spent some time a while back trying to get it to work with constexpr functions, but the main issue I kept running in to was how to get a dynamically sized array that works with constexpr. I suppose I could have a statically sized array and just have to keep it in sync with the number of entries.

I can take a look at it again and see if I can come up with something. If there's any prior art I could follow please let me know.

Reply:
I wasn't able to come up with a way to embed them at compile time like I would like, but I at least got rid of the allocations. Now there is an array and at runtime the values are assigned to the array without allocation.